tomcat-8.5/src/org/dogtagpki/tomcat/JSSUtil.java

```diff
@@ -19,38 +19,92 @@
 
 package org.dogtagpki.tomcat;
 
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
+
 import javax.net.ssl.KeyManager;
 import javax.net.ssl.TrustManager;
 
+import org.apache.juli.logging.Log;
+import org.apache.juli.logging.LogFactory;
 import org.apache.tomcat.util.net.SSLHostConfigCertificate;
 import org.apache.tomcat.util.net.jsse.JSSEKeyManager;
-import org.apache.tomcat.util.net.jsse.JSSEUtil;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import org.apache.tomcat.util.net.SSLContext;
+import org.apache.tomcat.util.net.SSLUtil;
+import org.apache.tomcat.util.net.SSLUtilBase;
 
+import org.mozilla.jss.crypto.Policy;
 import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
-import org.mozilla.jss.provider.javax.crypto.JSSTrustManager;
+import org.mozilla.jss.provider.javax.crypto.JSSNativeTrustManager;
+import org.mozilla.jss.ssl.SSLCipher;
+import org.mozilla.jss.ssl.SSLVersion;
 
-public class JSSUtil extends JSSEUtil {
+public class JSSUtil extends SSLUtilBase {
+    public static Log logger = LogFactory.getLog(JSSUtil.class);
 
-    public static Logger logger = LoggerFactory.getLogger(JSSUtil.class);
+    private String keyAlias;
 
     public JSSUtil(SSLHostConfigCertificate cert) {
         super(cert);
+
+        keyAlias = certificate.getCertificateKeyAlias();
         logger.debug("JSSUtil: instance created");
     }
 
     @Override
     public KeyManager[] getKeyManagers() throws Exception {
         logger.debug("JSSUtil: getKeyManagers()");
-        String keyAlias = certificate.getCertificateKeyAlias();
-        KeyManager keyManager = new JSSEKeyManager(new JSSKeyManager(), keyAlias);
-        return new KeyManager[] { keyManager };
+        return new KeyManager[] { new JSSKeyManager() };
     }
 
     @Override
     public TrustManager[] getTrustManagers() throws Exception {
         logger.debug("JSSUtil: getTrustManagers()");
-        return new TrustManager[] { new JSSTrustManager() };
+        return new TrustManager[] { new JSSNativeTrustManager() };
```

Comment:
Hm? I'm confused what you're asking. Note that Tomcat doesn't actually use the JSSTrustManager; it just feeds it to the SSLEngine.

`JSSNativeTrustManager` is a special trust manager. It is a [sentinel](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java#L349-L359) for adding a hook to use JSS's default cert validation logic that we used in [the old SSLSocket](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/ssl/common.c#L214-L216) code. This is the most performant `X509TrustManager` implementation: we add the callback handler, check the cert in C, and return the result. Other `X509TrustManager`s, in the future when they're supported, have to add a callback handler, use JNI calls to execute the `X509TrustManager` checking logic (which in turn, might execute more JNI calls if they call into the `CryptoManager`!), interpret the resulting Java exception, clear it, and set the appropriate return type. JNI calls are slow. 

You're right that it doesn't implement anything, but at the same time, it does exactly what it needs to, for the purposes of this code right now. :-)

Reply:
Yes, there's no question about using native code for trust manager for performance reasons. I was just asking about the use of `JSSNativeTrustManager` class.

IIUC the trust manager class is being checked here:
https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java#L350
so if it's a `JSSNativeTrustManager` it will use native trust manager instead of Java. The code inside the Java trust manager class itself is irrelevant.

I'm just wondering if we can accomplish the same thing without using a dummy class, for example, we can store a flag in `JSSTrustManager` that can be checked later by `JSSEngineReferenceImpl`:
```
return new TrustManager[] { new JSSTrustManager(JSSTrustManager.NATIVE) };
```
or use native trust manager by default in `JSSEngineReferenceImpl`:
```
private void applyTrustManagers() {

    if (trust_managers == null || trust_managers.length == 0) {
        // use native trust manager
        if (SSL.ConfigJSSDefaultCertAuthCallback(ssl_fd) == SSL.SECFailure) {
            throw new RuntimeException("Unable to configure trust manager: " + errorText(PR.GetError()));
        }
    } else {
        // use Java trust manager
    }
}
```
The JSS API might look cleaner that way.

Reply:
I think this is a fine concern, but I think we need to think about this more and see what other `TrustManagers` do. This is also mostly a JSS concern, so I think its fine to merge this without addressing this now. Worst case, we can always update TomcatJSS later. :-)